ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,427 @@
+import           Control.Monad hiding (ap)
+import           Control.Monad.Trans.Class (lift)
+import           Control.Monad.Trans.State.Strict hiding (state)
+import qualified Data.FingerTree.Strict as FT
+import           Data.Foldable
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (catMaybes, isJust)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           GHC.Generics (Generic)
+import           GHC.Show (showCommaSpace, showSpace)
+import           NoThunks.Class (NoThunks)
+import           Ouroboros.Consensus.Config.SecurityParam (SecurityParam (..))
+import           Ouroboros.Consensus.Ledger.Basics hiding (LedgerState)
+import           Ouroboros.Consensus.Storage.LedgerDB.HD
+import qualified Ouroboros.Network.AnchoredSeq as AS
+import           Ouroboros.Network.Block (HeaderHash, Point (..), SlotNo (..),
+                     StandardHash, castPoint, pattern GenesisPoint)
+import qualified Ouroboros.Network.Point as Point
+import           Test.Ouroboros.Storage.LedgerDB.OrphanArbitrary ()
+import           Test.QuickCheck
+import           Test.Tasty (TestTree, testGroup)
+import           Test.Tasty.QuickCheck (testProperty)
+import           Text.Show.Pretty (ppShow)
+
+
+tests :: TestTree
+tests = testGroup "Ledger" [ testGroup "DbChangelog"
+      [ testProperty "empty changelog satisfies invariants"
+        prop_emptySatisfiesInvariants
+      , testProperty "constructor generated changelog satisfies invariants"
+        prop_generatedSatisfiesInvariants
+      , testProperty "flushing keeps invariants"
+        prop_flushDbChangelogKeepsInvariants
+      , testProperty "rolling back keeps invariants"
+        prop_rollbackDbChangelogKeepsInvariants
+      , testProperty "prefixing back to anchor keeps invariants"
+        prop_prefixBackToAnchorKeepsInvariants
+      , testProperty "flushing keeps immutable tip"
+        prop_flushingKeepsImmutableTip
+      , testProperty "extending adds head to volatile states"
+        prop_extendingAdvancesTipOfVolatileStates
+      , testProperty "rollback after extension is noop"
+        prop_rollbackAfterExtendIsNoop
+      , testProperty "pruning leaves at most maxRollback volatile states"
+        prop_pruningLeavesAtMostMaxRollbacksVolatileStates
+      , testProperty "flushing splits immutable and volatile"
+        prop_flushingSplitsImmutableAndVolatile
+      , testProperty "prefixing back to anchor is rolling back volatile states"
+        prop_prefixBackToAnchorIsRollingBackVolatileStates
+      , testProperty "prefix back to volatile tip is a noop"
+        prop_rollBackToVolatileTipIsNoop
+      ]
+  ]
+
+{-------------------------------------------------------------------------------
+  Test setup
+-------------------------------------------------------------------------------}
+
+data TestLedger (mk :: MapKind) = TestLedger {
+  tlUtxos :: ApplyMapKind mk Key Int,
```

Comment:
I would comment why is it okay to represent this as a map from `Key` to `Int`. Will we always be increasing the `Int` or how would this work?

Also `ApplyMapKind mk = mk` so:
```suggestion
  tlUtxos :: mk Key Int,
```